include/boost/random/niederreiter_base2.hpp

```diff
@@ -0,0 +1,371 @@
+                                            , IntType
+                                            >
+{
+  typedef niederreiter_base2_engine<IntType, Nb2Table> self_t;
+  typedef detail::niederreiter_base2_lattice<IntType, Nb2Table> lattice_t;
+  typedef detail::gray_coded_qrng_base<self_t, lattice_t, IntType> base_t;
+
+public:
+  typedef typename base_t::result_type result_type;
+
+  //!Returns: Tight lower bound on the set of values returned by operator().
+  //!
+  //!Throws: nothing.
+  static BOOST_CONSTEXPR result_type min BOOST_PREVENT_MACRO_SUBSTITUTION ()
+  { return 0; }
+
+  //!Returns: Tight upper bound on the set of values returned by operator().
+  //!
+  //!Throws: nothing.
+  static BOOST_CONSTEXPR result_type max BOOST_PREVENT_MACRO_SUBSTITUTION ()
+  { return (std::numeric_limits<result_type>::max)(); }
+
+  //!Effects: Constructs the default `s`-dimensional Niederreiter base 2 quasi-random number generator.
+  //!
+  //!Throws: bad_alloc, invalid_argument, range_error.
+  explicit niederreiter_base2_engine(std::size_t s)
+    : base_t(s) // initialize lattice here
+  {}
+
+#ifdef BOOST_RANDOM_DOXYGEN
+  //=========================Doxygen needs this!==============================
+
+  //!Returns: The dimension of of the quasi-random domain.
+  //!
+  //!Throws: nothing.
+  std::size_t dimension() const { return base_t::dimension(); }
+
+  //!Requirements: *this is mutable.
+  //!
+  //!Effects: Resets the quasi-random number generator state to
+  //!the one given by the default construction. Equivalent to u.seed(0).
+  //!
+  //!\brief Throws: nothing.
+  void seed()
+  {
+    base_t::seed();
+  }
+
+  //!Requirements: *this is mutable.
+  //!
+  //!Effects: Effectively sets the quasi-random number generator state to the `init`-th
+  //!vector in the `s`-dimensional quasi-random domain, where `s` == X::dimension().
+  //!\code
+  //!X u, v;
+  //!for(int i = 0; i < N; ++i)
+  //!    for( std::size_t j = 0; j < u.dimension(); ++j )
+  //!        u();
+  //!v.seed(N);
+  //!assert(u() == v());
+  //!\endcode
+  //!
+  //!\brief Throws: overflow_error.
+  void seed(IntType init)
+  {
+    base_t::seed(init);
+  }
+
+  //!Requirements: *this is mutable.
+  //!
+  //!Returns: Returns a successive element of an `s`-dimensional
+  //!(s = X::dimension()) vector at each invocation. When all elements are
+  //!exhausted, X::operator() begins anew with the starting element of a
+  //!subsequent `s`-dimensional vector.
+  //!
+  //!Throws: overflow_error.
+  result_type operator()()
+  {
+    return base_t::operator()();
+  }
+
+  //!Requirements: *this is mutable.
+  //!
+  //!Effects: Advances *this state as if `z` consecutive
+  //!X::operator() invocations were executed.
+  //!\code
+  //!X u = v;
+  //!for(int i = 0; i < N; ++i)
+  //!    u();
+  //!v.discard(N);
+  //!assert(u() == v());
+  //!\endcode
+  //!
+  //!Throws: overflow_error.
+  void discard(IntType z)
```

Comment:
IntType is wrong.  It should always be unsigned long long (boost::uinmax_t).  Consider the case where IntType = uint8_t, with 2 dimensions.  The generator can produce 256 vectors, which is 512 elements, which is more than IntType can represent.